subworkflows/local/bamfiltering.nf

```diff
@@ -80,45 +80,40 @@ workflow FILTER_BAM {
     }
 
     if ( ( params.run_metagenomics && params.metagenomics_input == 'unmapped' ) && params.preprocessing_skippairmerging ) {
-        ch_paired_fastq_for_cat = SAMTOOLS_FASTQ_UNMAPPED.out.fastq
+        ch_paired_fastq_for_cat = SAMTOOLS_FASTQ_UNMAPPED.out.fastq.filter { !it[0].single_end }
+
+        ch_single_fastq_for_cat = SAMTOOLS_FASTQ_UNMAPPED.out.fastq
                                     .mix(SAMTOOLS_FASTQ_UNMAPPED.out.singleton)
                                     .mix(SAMTOOLS_FASTQ_UNMAPPED.out.other)
                                     .groupTuple()
-                                    .map {
-                                        meta, fastqs ->
-                                            def meta_new = meta.clone()
-                                            meta_new['single_end'] = true
-                                        [ meta_new, fastqs.flatten() ]
-                                    }
-        CAT_FASTQ_UNMAPPED ( ch_paired_fastq_for_cat )
+                                    .filter{ it[0].single_end }
+
+        CAT_FASTQ_UNMAPPED ( ch_single_fastq_for_cat )
     }
 
-    // TODO: see request https://github.com/nf-core/eager/issues/945
     if ( ( params.run_metagenomics && ( params.metagenomics_input == 'mapped' || params.metagenomics_input == 'all' ) ) && params.preprocessing_skippairmerging ) {
-        ch_paired_fastq_for_cat = SAMTOOLS_FASTQ_MAPPED.out.fastq
+        ch_paired_fastq_for_cat = SAMTOOLS_FASTQ_MAPPED.out.fastq.filter { !it[0].single_end }
+
+        ch_single_fastq_for_cat = SAMTOOLS_FASTQ_MAPPED.out.fastq
                                     .mix(SAMTOOLS_FASTQ_MAPPED.out.singleton)
                                     .mix(SAMTOOLS_FASTQ_MAPPED.out.other)
                                     .groupTuple()
-                                    .map {
-                                        meta, fastqs ->
-                                            def meta_new = meta.clone()
-                                            meta_new['single_end'] = true
-                                        [ meta_new, fastqs.flatten() ]
-                                    }
-        CAT_FASTQ_MAPPED ( ch_paired_fastq_for_cat )
+                                    .filter{ it[0].single_end }
+
+        CAT_FASTQ_MAPPED ( ch_single_fastq_for_cat )
     }
 
     // Routing for metagenomic screening -> first accounting for paired-end mapping, then merged mapping, then no metagenomics
     if ( ( params.run_metagenomics && params.metagenomics_input == 'unmapped' ) && params.preprocessing_skippairmerging ) {
-        ch_fastq_for_metagenomics = CAT_FASTQ_UNMAPPED.out.reads
+        ch_fastq_for_metagenomics = CAT_FASTQ_UNMAPPED.out.reads.mix(ch_paired_fastq_for_cat)
     } else if ( ( params.run_metagenomics && ( params.metagenomics_input == 'mapped' || params.metagenomics_input == 'all' ) ) && params.preprocessing_skippairmerging ) {
```

Comment:
if input now is 'all', the unmapped reads are not taken into account.

I think that needs one more if-condition, with something like 
if metagenomics_input == 'all':
CAT_FASTQ_MAPPED.out.reads.mix(CAT_FASTQ_UNMAPPED.out.reads).mix(ch_paired_fastq_for_cat)

but I might miss something here?

Reply:
It looks like the unmapped reads are not filtered out? then it makes sense...

Reply:
the splitting of all vs only mapped reads is dealt with at the modules.config level. 

`    withName: SAMTOOLS_FASTQ_MAPPED {
        tag = { "${meta.reference}|${meta.sample_id}_${meta.library_id}" }
        ext.args = [
            params.metagenomics_input == 'all' ? '' : '-F 4',
`

We could split the module/import into an additional SAMTOOLS_FASTQ_ALL, for clarity for anyone looking into this section of the pipeline... @jfy133 what are your thoughts?

Reply:
I adjusted this whole section to imporve clarity. please see latest commit.

essentially the splitting of the fastq and cat into unmapped and mapped versions is unnecessary. all parameters can be parsed fine within modules config.

i confirmed that the md5sum of the final files sent to the ch_fastq_for_metagenomics channel are consistent across all the parameter combos that we were there.